lte/gateway/c/session_manager/UEShard.cpp

```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright 2020 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <string>
+#include <utility>
+#include <vector>
+#include <algorithm>
+#include <map>
+
+#include "UEShard.h"
+#define MAX_SHARD_SIZE 100
+
+namespace magma {
+
+UEShard::UEShard() {
+  number_of_shards = 0;
+}
+
+// returns shard_id associated with UE added
+int UEShard::add_ue(std::string imsi) {
+  int shard_id;
+  if (shards.size() == 0) {
+    shard_id = 1;
+    shards[shard_id].push_back(imsi);
+    number_of_shards++;
+  } else {
+    // iterate through all shards to find open location to put imsi
+    for (auto it = shards.begin(); it != shards.end(); it++) {
+      if (total_ues_for_shard(it->first) < MAX_SHARD_SIZE) {
+        shard_id = it->first;
+        shards[shard_id].push_back(imsi);
+        return shard_id;
+      }
+    }
+    shard_id = ++number_of_shards;
+    shards[shard_id].push_back(imsi);
+  }
+  return shard_id;
+}
+
```

Comment:
Shards are normally used to spread the load among them. Per what I understand, this algorithm will fill a shard completely before moving to the next shard. So that means, if you have 100 UEs, all UEs will fall in the same shard and all the load will be on that shard. Is this intentional?

If not, maybe a good way to distribute the load is to use a uniform hash like MD5 of the IMSI (or any other uniform distributed algorithm that doesn't take a hit on performance)

Reply:
It's intentional. The idea according to Nick was to have a fixed size for each shard(like 100), fill up that shard for insertion, and once it's full to go to the next shard. I think right now it's just meant to be simple and I'll incorporate rebalancing down the line

Reply:
@koolzz probably you already did it the past but can you
- Provide more insgith on the need of this sharing?
- Will sessiond send a list of all the UEs in the shard every time we request update for that shard? Or will pipelined know what UEs are included on that shard?
- Why can't we just poll those subcribers independently without having to shard based on current state of the quota (if over 80% we hyper poll them)

Reply:
Answers from Pravin: 
1. sessionD sends shard ID to pipelineD to retrieve stats for all UEs in the shard
2. the trigger is a request for updated stats for any UE in the shard(to differentiate among multiple shards)
3. querying each individual UE generates too much overhead
The reason to introduce sharding as solution is to scale UP the system, with multiple shards you can process each shard independently. sessionD does not need multiple threads, but pipelineD can make use of multiple requests for using async IO more effectively

Reply:
Thanks. 

I think a change like this that increases complexity of the system requires a proper reason and  design document showing the call flow and what is the intention of this algorithm on the long term.

I think that before starting any coding a proper load balance strategy must be decided. If there is no clear strategy, sharding will become as bad as querying all UEs

Reply:
I think this feature is documented in google doc, @veshkemburu can you post a PR for design doc for this feature in markdown format.

Reply:
Please, add on that document some diagrams of the call flow.

I think it would be good to put in a diagram how you expect the system behave (with load balancing included, even you don't have the algo)

Reply:
I'll have the design doc pushed in a separate PR.